### PR TITLE
ContentNegotiation feature for client

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-1689

Extract `ContentNegatiaion` feature into common module to share it between client and server.
* All serializers are shared now.
* `JsonFeature` was removed in favour of new client-side `ContentNegatitaion`
* Deleted depraceted API, related to this change

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -101,3 +95,8 @@ include ':ktor-network:ktor-network-tls'
 include ':ktor-network:ktor-network-tls:ktor-network-tls-certificates'
 include ':ktor-bom'
 include ':ktor-test-dispatcher'
+include ':ktor-common'
```
Comment (e5l) [inline settings.gradle:98]: `ktor-common` is really confusing name here. I can't even guess what should be there

**Review by e5l — CHANGES_REQUESTED** [review]
The PR is really hard to review cause there are many changes inside:
- type info refactoring
- extraction of common in serialization
- utils refactoring

Could you split the PR to make changes gradually?

File: settings.gradle (hunk 1 above)
Comment (rsinukov) [reply settings.gradle:98]: This is for sharing code between server and client, and it's too small to be in a top level module. Do you have a better naming ideas?

**Comment by rsinukov** [comment]
@e5l there were some not relevant changes after rebasing 2.0.0. Now this PR is smaller. I can try to separate it into different PRs, but since most of the changes are in tests and most of the new code is just moving of an old feature, it's hard to split to make it cosiderably smaller.

File: settings.gradle (hunk 1 above)
Comment (e5l) [reply settings.gradle:98]: Let's discuss how are we going to have such modules. I'm not sure that we should put all common between all features in a single common module. Maybe it's better to have ktor-serialization-common

Comment (e5l) [reply settings.gradle:98]: btw we have a name conflict here: there are 2 ktor-serialization modules(cause all modules start being plain during publishing)

**Comment by rsinukov** [comment]
@e5l I split this PR into three parts. Please review #2441, #2442 and #2443

**Pull request closed without merging**